extra_geom/base.py

```diff
@@ -794,6 +794,77 @@ def data_coords_to_positions(self, module_no, slow_scan, fast_scan):
             + (np.expand_dims(tile_ss, -1) * coords_ss_vec) \
             + (np.expand_dims(tile_fs, -1) * coords_fs_vec)
 
+    def positions_to_data_coords(self, positions,  origin=None, unit=None):
+        """Convert assembled positions to data array coordinates
+
+        Data array coordinates are how you might refer to a pixel in an array of
+        detector data: module number, and indices in the slow-scan and fast-scan
+        directions. But coordinates in the two pixel dimensions aren't necessarily
+        integers, e.g. if they refer to the centre of a peak.
+
+        Positions is a 2D array (n, 2) describing (x, y) positions of points within an
+        assembled detector image. The may hold floats for sub-pixel positions.
+
+        If origin is None (default) the position indices are relative to the centre of
+        the geometry. `origin=(0, 0)` would set the origin to the first index of the
+        assembled array.
+        If unit is None (default) the positions given are in pixel, else given positions
+        are in meter relative to the centre, e.g. `unit=1e-3` interprets positions in mm.
+
+        .. note::
+
+          This uses an assembly limited to snapping pixel to a regular grid, so any
+          subpixel offset of tile or rotation in your geometry will be ignored.
+
+        Returns 4 array of similar length for module, slow scan and fast scan
+        coordinates and a sub-array of the input data excluding invalid coordinates.
+        note: invalid coordinates are coordinates either outside the assembled image, or
+        between detector tiles.
+        """
+        # TODO what coordinates as input: (x, y) or (y, x)?
+        positions = np.asarray(positions)
+
+        unit = unit or self.pixel_size
+        if origin is None:
+            origin = self._snapped().centre[::-1]
+        else:
+            origin = np.asarray(origin) * unit / self.pixel_size
+
+        pos = np.rint(np.asarray(positions) * unit / self.pixel_size + origin).astype(int)
+
+        xx, yy = np.meshgrid(
+            np.arange(self.expected_data_shape[-1]),
+            np.arange(np.prod(self.expected_data_shape[:-1])),
+            indexing='xy'
+        )
+        pxx = self.position_modules(xx.reshape(self.expected_data_shape).astype(float))[0]
+        pyy = self.position_modules(yy.reshape(self.expected_data_shape).astype(float))[0]
```

Comment:
If I'm following correctly, these 'x' and 'y' things should really be 'fast scan' and 'slow scan'. E.g. for AGIPD, the fast scan direction, which you've called 'x' here, is typically along the y axis.

(I know I sound like a broken record about this, but geometry code is hard to follow at the best of times, so I really want to avoid making it harder with half-right names)

Reply:
These are the x and y coordinates in a (2D) data array. I can call them fast and slow scan if you prefer but I found it more confusing as it is not the same as what we call ss/fs usually in this code base. And in this case x will always be fast scan and y slow scan.